### PR TITLE
Added the popular GTZAN dataset:

### PR DESCRIPTION
**Pull request by mmxgn**
* Added the GTZAN class in torchaudio.datasets using the same format as the rest of the datasets.
* Added the appropriate test function in test_datasets.py.
* Added the GTZAN class in the datasets.rst documentation file.

### CODE REVIEW

**Comment by mmxgn** [comment]
The GTZAN is a very popular dataset for Genre classification first described in:
```" Musical genre classification of audio signals " by G. Tzanetakis and P. Cook in IEEE Transactions on Audio and Speech Processing 2002.```

While its popularity might have fallen recently due to larger and better datasets I believe it is still an essential dataset to have due to its small size and its ubiquity in the literature and otherwise. Lots of tutorials on the web make use of it.

Example of usage:
```python
FFT_HOP = 256
FFT_SIZE = 512
N_MELS = 96

gtzan_ds = GTZAN('data', 
                 filtered='training', 
                 download=True,
                 transform=torchaudio.transforms.MelSpectrogram(
                 n_fft=FFT_SIZE,
                 hop_length=FFT_HOP,
                 n_mels=N_MELS)
                )

gtzan_ds[0]
```

Output:

```python
(tensor([[[2.4445e-01, 1.1175e-02, 2.2179e-02,  ..., 3.6174e-02,
           2.0766e-01, 4.3936e+00],
          [5.1548e-01, 2.3564e-02, 4.6768e-02,  ..., 7.6280e-02,
           4.3789e-01, 9.2648e+00],
          [4.2430e-02, 3.6839e+00, 4.0256e+00,  ..., 7.7146e+00,
           4.7242e-01, 7.4579e+00],
          ...,
          [1.4867e-03, 2.7238e-07, 2.4675e-06,  ..., 1.7559e-06,
           1.2393e-06, 5.9337e-03],
          [1.3977e-03, 1.1830e-07, 1.3188e-07,  ..., 1.8985e-07,
           2.2827e-07, 6.1415e-03],
          [1.3772e-03, 1.3670e-07, 9.7792e-08,  ..., 1.2337e-07,
           1.6914e-07, 6.3171e-03]]]),
 22050,
 'blues')

Additionally, since the original dataset does not provide a train,test split, I took the one from 
```
https://github.com/jordipons/sklearn-audio-transfer-learning/tree/master/data/index/GTZAN
```
Which mitigates some (all?) of the duplication issues in the GTZAN and allows for comparison with the method available in this repository. This split can be enabled with the `filtered=True` initialization keyword.

File: torchaudio/datasets/gtzan.py [hunk 1]
```diff
@@ -0,0 +1,1064 @@
...
+# The following lists prefixed with `filtered_` provide a filtered split
+# that:
+#
+# a. Mitigate a known issue with GTZAN (duplication)
+#
+# b. Provide a standard split for testing it against other
+#    methods (e.g. the one in jordipons/sklearn-audio-transfer-learning).
+#
+# Those are used when GTZAN is initialised with the `filtered` keyword.
+# The split was taken from (github) jordipons/sklearn-audio-transfer-learning.
```
Comment (vincentqb) [inline torchaudio/datasets/gtzan.py:22]: Are there other splits that people use instead?

File: test/test_datasets.py [hunk 2]
```diff
@@ -55,6 +56,11 @@ def test_speechcommands(self):
         data = SPEECHCOMMANDS(self.path)
         data[0]
 
+    def test_gtzan(self):
+        data = GTZAN(self.path)
```
[the remark this answers is not in the document]
Comment (mmxgn) [reply test/test_datasets.py:60]: I am sorry it seems I missed this part. I will fix it.

File: torchaudio/datasets/gtzan.py [hunk 3]
```diff
@@ -0,0 +1,1064 @@
...
+    _ext_audio = ".wav"
+
+    def __init__(
+        self,
+        root: str,
+        url: str = URL,
+        folder_in_archive: str = FOLDER_IN_ARCHIVE,
+        download: bool = False,
+        filtered: bool = False,
+        subset: str = "",
```
[the remark this answers is not in the document]
Comment (mmxgn) [reply torchaudio/datasets/gtzan.py:1001]: This is interesting. I thought I had removed it. I will be more careful in the future. I will fix it for now.

File: torchaudio/datasets/gtzan.py [hunk 4]
```diff
@@ -0,0 +1,1064 @@
...
+        transform: Any = None,
+        target_transform: Any = None,
```
[the remark this answers is not in the document]
Comment (mmxgn) [reply torchaudio/datasets/gtzan.py:1003]: I am sorry about that. I will remove them. 

I had started working on the dataset looking at the YESNO dataset as a template which has the transforms. Is that intentional? If not, is it a bug? Should I remove the transforms from there while I am at it?

File: torchaudio/datasets/gtzan.py [hunk 5]
```diff
@@ -0,0 +1,1064 @@
...
+        self.subset = subset
+        self.transform = transform
+        self.target_transform = target_transform
+
+        assert (
+            not filtered or subset in ["training", "validation", "testing"] and filtered
+        ), (
+            "When `filtered` is True, subset must take a value from "
+            + "{'training', 'validation', 'testing'}, otherwise `filtered` must be False."
+        )
```
[the remark this answers is not in the document]
Comment (mmxgn) [reply torchaudio/datasets/gtzan.py:1021]: You are correct. I will get to it.

File: torchaudio/datasets/gtzan.py (hunk 1 above)
Comment (mmxgn) [reply torchaudio/datasets/gtzan.py:22]: AFAIK yes but this one has been used in a previous ISMIR late breaking demo by Jordi Pons and Xavier Serra:

```Jordi Pons, Xavier Serra.  “musicnn:  pre-trained convolutional neural networks for music audio tagging”, Late Break-ing/Demo at the 20th International Society for Music Information Retrieval, Delft, The Netherlands, 2019.```

As well as their transfer learning with MusiCNN:
https://github.com/jordipons/sklearn-audio-transfer-learning/

**Comment by mmxgn** [comment]
Hi,

Thanks for being patient with me :)

Some tests failed for reasons unknown when running curl while installing:
```
Fetching liblame from https://downloads.sourceforge.net/project/lame/lame/3.99/lame-3.99.5.tar.gz
curl: (7) Failed to connect to 2607:f748:10:12::5f:2: Cannot assign requested address
```

Is there a way to rerun the failed builds?

File: torchaudio/datasets/gtzan.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vincentqb) [reply torchaudio/datasets/gtzan.py]: In YESNO, those parameters will give a deprecation warning in `__init__` :)

**Comment by vincentqb** [comment]
> Thanks for being patient with me :)

And thanks for your contributions! :)

> Some tests failed for reasons unknown when running curl while installing:
> 
> ```
> Fetching liblame from https://downloads.sourceforge.net/project/lame/lame/3.99/lame-3.99.5.tar.gz
> curl: (7) Failed to connect to 2607:f748:10:12::5f:2: Cannot assign requested address
> ```
> 
> Is there a way to rerun the failed builds?

Unfortunately, we do not have a "bot" to trigger the build. However, as noted in pytorch/pytorch#17057, you can locally run `git commit --amend` and then force push to the branch. This will change the hash of you latest commit (without changing the content or message), and trigger the tests again.

**Pull request merged**